projects/charger/src/charger_controller.c

```diff
@@ -0,0 +1,127 @@
+#include "charger_controller.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "can_interval.h"
+#include "can_transmit.h"
+#include "charger_events.h"
+#include "event_queue.h"
+#include "generic_can.h"
+#include "generic_can_msg.h"
+#include "status.h"
+
+#define CHARGER_PERIOD_US 1000000  // 1 Second as defined in datasheet.
+
+#define CHARGER_EXPECTED_RX_DLC 5
+#define CHARGER_EXPECTED_RX_ID 0x18FF50E5
+
+#define CHARGER_EXPECTED_TX_DLC 5
+#define CHARGER_EXPECTED_TX_ID 0x1806E5F4
+
+static CanInterval *s_interval;
+static ChargerStatus *s_charger_status;
+static GenericCanMsg s_tx_msg;
+
+typedef union J1939CanId {
+  uint32_t raw_id;
+  struct {
+    uint32_t source_address : 8;  // Source
+    uint32_t pdu_specifics : 8;   // Destination
+    uint32_t pdu_format : 8;      // Packet Format
+    uint32_t dp : 1;              // Always 0
+    uint32_t r : 1;               // Always 0
+    uint32_t priority : 3;        // Anything
+  };
+} J1939CanId;
+
+// Explicit for readability.
+static const J1939CanId s_rx_id = {
+  .source_address = 0xE5,  // Indicates BMS in the J1939 standard.
+  .pdu_specifics = 0x50,   // Broadcast address (BCA) id.
+  .pdu_format = 0xFF,      // From datasheet.
+  .dp = 0,
+  .r = 0,
+  .priority = 0x06,  // From datasheet.
+};
+static const J1939CanId s_tx_id = {
+  .source_address = 0xF4,  // Indicates BMS in the J1939 standard.
+  .pdu_specifics = 0xE5,   // Charger control system (CCS) id.
+  .pdu_format = 0x06,      // From datasheet.
+  .dp = 0,
+  .r = 0,
+  .priority = 0x06,  // From datasheet.
+};
+
+// GenericCanRx
+static void prv_rx_handler(const GenericCanMsg *msg, void *context) {
+  (void)context;
```

Comment:
From speaking with @karlding my understanding is that the CAN UART slave board is located between the charger and the interface board so this would actually use CAN UART not CAN HW. I just use HW in order to allow the tests to pass on Travis since there is no loopback for UART and HW supports extended. However, if my understanding is backwards then yes you would be correct.

```
---------------                   ------------------                ------------------------               -----------
| REST OF CAR | <- NETWORK CAN -> | CHARGER MASTER | <- CAN UART -> | GENERIC CAN UART SLAVE| <- CAN HW -> | CHARGER |
---------------                   ------------------                ------------------------               -----------
```

Reply:
To be specific, I mean the `context` variable. It isn't actually used anywhere yet but you have a TODO about passing data through to system CAN.

Reply:
Oh I was actually planning to just use `GenericCanNetwork` or rely on the direct translation code I wrote in `generic_can_msg.h` and then just use `can_transmit`? I didn't implement this since it requires another roll of codegen tooling to define the message and I wanted to wait to make sure there weren't more changes that needed to be made.

Perhaps my wording in the test is unclear? I do use HW exclusively in the test because using the Network CAN would conflict with initializing GenericCanHw.

Reply:
Ah, my bad. I misspoke. I saw that you had `GenericCan *can` in your settings struct and meant CAN network, not HW.

Reply:
Yeah this is likely the only to use of it unless I make the `notify` (previously permissions`) module use a `CanInterval` to send the notification periodically as a sort of heartbeat.